### PR TITLE
Allow parameters in module compose

### PR DESCRIPTION
**Pull request by samedii**

### CODE REVIEW

File: workflow/torch/module_compose.py [hunk 1]
```diff
@@ -20,6 +20,20 @@ def __init__(self, *modules_and_functions):
...
+        self.parameter_list = nn.ParameterList([
+            parameter
+            for parameter in self.modules_and_functions
+            if isinstance(parameter, nn.Parameter)
+        ] + [
+            parameter_or_tuple[0]
+            for parameter_or_tuple in self.modules_and_functions
+            if (
+                type(parameter_or_tuple) is tuple and
+                isinstance(parameter_or_tuple[0], nn.Parameter)
```
Comment (Jim-Holmstroem) [inline workflow/torch/module_compose.py:32]: I'm not super well versed how it works in practice, does all elements of the tuple need to be a nn.Parameter?

Comment (samedii) [reply workflow/torch/module_compose.py:32]: No only the first should be, the second should be a function

Comment (Jim-Holmstroem) [reply workflow/torch/module_compose.py:32]: aah, wait, forgot that you're pushing it into nn.ParameterList, you can divert the responsibility to not have a corrupt state to that

**Comment by FelixAbrahamsson** [comment]
Can you explain how this is used? I'm guessing you want to input parameters in such a way that they are registered by autograd. So it would be something like
```
ModuleCompose(
    some_fn1,
    some_fn2,
    some_parameter1,
    some_parameter2,
)
```
If so, wouldn't it be nicer to input a list of parameters as an explicit argument? Something like
`ModuleCompose(*modules_and_functions, parameters=[])`

**Comment by samedii** [comment]
> Can you explain how this is used? I'm guessing you want to input parameters in such a way that they are registered by autograd. So it would be something like
> 
> ```
> ModuleCompose(
>     some_fn1,
>     some_fn2,
>     some_parameter1,
>     some_parameter2,
> )
> ```
> 
> If so, wouldn't it be nicer to input a list of parameters as an explicit argument? Something like
> `ModuleCompose(*modules_and_functions, parameters=[])`

It works just like we do with nn.Module already. It's just that nn.Parameter is not a nn.Module for some reason.
```
ModuleCompose(
    ...
    (nn.Conv2d(5, 5, kernel_size=5, padding=2), lambda module, x: module(x) + x),
    (nn.Parameter(torch.randn(1, 5, 1, 1), requires_grad=True), lambda bias, x: x + bias),
)
```

**Comment by FelixAbrahamsson** [comment]
Alright, a `torch.nn.Parameter` doesn't have a `__call__` method though so I think we would need to limit it to only allow inputs of the format `(Parameter, function)`?

**Comment by samedii** [comment]
> Alright, a `torch.nn.Parameter` doesn't have a `__call__` method though so I think we would need to limit it to only allow inputs of the format `(Parameter, function)`?

I am a little confused. Do you want to do additional checks that we didn't have for `nn.Module`?

**Comment by FelixAbrahamsson** [comment]
> I am a little confused. Do you want to do additional checks that we didn't have for `nn.Module`?

Unless I'm mistaken, `.forward` will now crash on lines 47 and 49
```
            else:
                if type(x) is tuple:
                    x = module_or_function(*x)
                else:
                    x = module_or_function(x)
```
if a `Parameter` is passed to `ModuleCompose` without a corresponding function? I guess we can allow that to crash if there is no use case for it.

**Comment by samedii** [comment]
I see. I don't think that a nn.Parameter should be allowed to be passed ModuleCompose on its own. Compose puts together functions. The tuples are just there because we need to take in state and save it for the optimizer.

**Comment by FelixAbrahamsson** [comment]
> I don't think that a nn.Parameter should be allowed to be passed ModuleCompose on its own. 

Do we want a check for it or just assume people don't do it?

**Comment by samedii** [comment]
We are not checking anything else at the moment and I think that we should retire ModuleCompose in favor of ModuleChain. We can have a discussion on how much we want to "protect against corrupt state" in slack

**Review by FelixAbrahamsson — APPROVED** [review]

**Pull request merged**